File: backend/infrastructure/sse.py
"""
Server-Sent Events (SSE) broadcaster for real-time streaming.

Manages per-room subscriber queues and broadcasts streaming events
from agent responses to connected SSE clients.
"""

import asyncio
import json
import logging
from collections import defaultdict

logger = logging.getLogger(__name__)


class EventBroadcaster:
    """
    Per-room SSE event broadcaster using asyncio.Queue per client.

    Each SSE client subscribes to a room and receives events via its own queue.
    Broadcasts are non-blocking (put_nowait) to avoid blocking agent generation.
    """

    def __init__(self, max_queue_size: int = 256):
        self._subscribers: dict[int, set[asyncio.Queue]] = defaultdict(set)
        self._max_queue_size = max_queue_size
        self._shutdown_event = asyncio.Event()

    def subscribe(self, room_id: int) -> asyncio.Queue:
        """Create a new subscriber queue for a room."""
        queue: asyncio.Queue = asyncio.Queue(maxsize=self._max_queue_size)
        self._subscribers[room_id].add(queue)
        logger.debug(f"SSE subscriber added for room {room_id} (total: {len(self._subscribers[room_id])})")
        return queue

    def unsubscribe(self, room_id: int, queue: asyncio.Queue) -> None:
        """Remove a subscriber queue from a room."""
        subs = self._subscribers.get(room_id)
        if subs:
            subs.discard(queue)
            if not subs:
                del self._subscribers[room_id]
        logger.debug(
            f"SSE subscriber removed for room {room_id} "
            f"(remaining: {len(self._subscribers.get(room_id, set()))})"
        )

    def broadcast(self, room_id: int, event: dict) -> None:
        """
        Broadcast an event to all subscribers of a room.

        Non-blocking: drops events if a subscriber's queue is full
        rather than blocking the agent generation pipeline.
        """
        subs = self._subscribers.get(room_id)
        if not subs:
            return

        data = json.dumps(event)
        for queue in subs:
            try:
                queue.put_nowait(data)
            except asyncio.QueueFull:
                logger.warning(f"SSE queue full for room {room_id}, dropping event type={event.get('type')}")

    def get_subscriber_count(self, room_id: int) -> int:
        """Get the number of active subscribers for a room."""
        return len(self._subscribers.get(room_id, set()))

    def has_subscribers(self, room_id: int) -> bool:
        """Check if a room has any active SSE subscribers."""
        return bool(self._subscribers.get(room_id))

    @property
    def shutdown_event(self) -> asyncio.Event:
        """Event that is set when the server is shutting down."""
        return self._shutdown_event

    def shutdown(self) -> None:
        """Signal all SSE connections to terminate."""
        self._shutdown_event.set()
        # Wake up all subscriber queues so they can exit
        for room_subs in self._subscribers.values():
            for queue in room_subs:
                try:
                    queue.put_nowait(None)
                except asyncio.QueueFull:
                    pass
        logger.info("SSE broadcaster shutdown signalled")

File: backend/src/infrastructure/sse.ts
/**
 * Per-room SSE fan-out. One queue per client; a broadcast is a non-blocking push
 * onto each, so a slow reader can never stall the agent generating events. On
 * overflow {@link AsyncQueue} drops the *oldest* event, deliberately: for a stream
 * of `content_delta`s the newest is the one that matters, and message events
 * survive either way since `usePolling` re-fetches them.
 */

import { AsyncQueue } from '@/lib/async-queue'
import { getLogger } from './logging/logger'

const logger = getLogger('SSE')

// Per client, not per room: a room with twenty viewers holds twenty of these.
const DEFAULT_MAX_QUEUE_SIZE = 256

/** `type` doubles as the SSE `event:` name — `useSSE.ts` attaches listeners by
 * it — so it is required, never defaulted. */
export interface SseEvent {
  type: string
  [key: string]: unknown
}

export type SseSubscription = AsyncQueue<string>

export class EventBroadcaster {
  private readonly subscribers = new Map<number, Set<SseSubscription>>()
  private shuttingDown = false

  constructor(private readonly maxQueueSize: number = DEFAULT_MAX_QUEUE_SIZE) {}

  get isShuttingDown(): boolean {
    return this.shuttingDown
  }

  /** The caller must {@link unsubscribe} it. */
  subscribe(roomId: number): SseSubscription {
    const queue = new AsyncQueue<string>(this.maxQueueSize)
    let room = this.subscribers.get(roomId)
    if (!room) {
      room = new Set()
      this.subscribers.set(roomId, room)
    }
    room.add(queue)
    logger.debug(`SSE subscriber added for room ${roomId} (total: ${room.size})`)
    return queue
  }

  /** Ending the queue as well as dropping it matters: a reader parked in `next()`
   * is only woken by `end()` or a push, so the loop would hang until keepalive. */
  unsubscribe(roomId: number, queue: SseSubscription): void {
    const room = this.subscribers.get(roomId)
    if (room) {
      room.delete(queue)
      if (room.size === 0) this.subscribers.delete(roomId)
    }
    queue.end()
    logger.debug(
      `SSE subscriber removed for room ${roomId} (remaining: ${this.getSubscriberCount(roomId)})`,
    )
  }

  /** A no-op when nobody is listening — the polling fallback needs no SSE. */
  broadcast(roomId: number, event: SseEvent): void {
    const room = this.subscribers.get(roomId)
    if (!room || room.size === 0) return

    let data: string
    try {
      data = JSON.stringify(event)
    } catch (error) {
      // A caller bug, but it must not take down the turn that produced it.
      logger.exception(`Dropping unserialisable SSE event type=${event.type}`, error)
      return
    }

    for (const queue of room) queue.push(data)
  }

  getSubscriberCount(roomId: number): number {
    return this.subscribers.get(roomId)?.size ?? 0
  }

  hasSubscribers(roomId: number): boolean {
    return this.getSubscriberCount(roomId) > 0
  }

  /** Ends the queues rather than only setting the flag, so parked readers wake at
   * once instead of after one full {@link KEEPALIVE_INTERVAL_MS} each. */
  shutdown(): void {
    this.shuttingDown = true
    for (const room of this.subscribers.values()) {
      for (const queue of room) queue.end()
    }
    logger.info('SSE broadcaster shutdown signalled')
  }
}

/** Silence between events before a comment frame is sent. */
export const KEEPALIVE_INTERVAL_MS = 15_000
